entproto/cmd/protoc-gen-entgrpc/main.go

```diff
@@ -41,52 +46,166 @@ func main() {
+// processFile generates service implementations from all services defined in the file.
+func processFile(gen *protogen.Plugin, file *protogen.File, graph *gen.Graph) error {
 	if len(file.Services) == 0 {
 		return nil
 	}
 	for _, s := range file.Services {
-		generateService(gen, file, graph, s)
+		sg, err := newServiceGenerator(gen, file, graph, s)
+		if err != nil {
+			return err
+		}
+		if err := sg.generate(); err != nil {
+			return err
+		}
 	}
 	return nil
 }
 
-// generateService creates a file named <package>_<service_name>.go with an implementation of a *protogen.Service.
-func generateService(gen *protogen.Plugin, file *protogen.File, graph *gen.Graph, s *protogen.Service) {
-	filename := file.GeneratedFilenamePrefix + "_" + snake(s.GoName) + ".go"
-	g := gen.NewGeneratedFile(filename, file.GoImportPath)
+func newServiceGenerator(plugin *protogen.Plugin, file *protogen.File, graph *gen.Graph, service *protogen.Service) (*serviceGenerator, error) {
+	adapter, err := entproto.LoadAdapter(graph)
+	if err != nil {
+		return nil, err
+	}
+	typeName, err := extractEntTypeName(service, graph)
+	if err != nil {
+		return nil, err
+	}
+	filename := file.GeneratedFilenamePrefix + "_" + snake(service.GoName) + ".go"
+	g := plugin.NewGeneratedFile(filename, file.GoImportPath)
+	fieldMap, err := adapter.FieldMap(typeName)
+	if err != nil {
+		return nil, err
+	}
+	return &serviceGenerator{
+		GeneratedFile: g,
+		entPackage:    protogen.GoImportPath(graph.Config.Package),
+		file:          file,
+		service:       service,
+		typeName:      typeName,
+		fieldMap:      fieldMap,
+	}, nil
+}
 
+func (g *serviceGenerator) generate() error {
 	g.P("// Code generated by protoc-gen-entgrpc. DO NOT EDIT.")
-	g.P("package ", file.GoPackageName)
+	g.P("package ", g.file.GoPackageName)
 	g.P()
-	entClient := protogen.GoImportPath(graph.Config.Package).Ident("Client")
-	g.P("// ", s.GoName, " implements ", s.GoName, "Server")
-	g.P("type ", s.GoName, " struct {")
-	g.P("client *", entClient)
-	g.P("Unimplemented", s.GoName, "Server")
-	g.P("}")
+	g.generateConstructor()
+	g.P()
+	if err := g.generateToProtoMapper(); err != nil {
+		return err
+	}
 	g.P()
-	g.P("func New", s.GoName, "(client *", entClient, ") *", s.GoName, "{")
-	g.P("return &", s.GoName, "{client: client}")
+
+	for _, method := range g.service.Methods {
+		if err := g.generateMethod(method); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (g *serviceGenerator) generateConstructor() {
+	g.Tmpl(`
+	// %(svcName) implements %(svcName)Server
+	type %(svcName) struct {
+		client *%(entClient)
+		Unimplemented%(svcName)Server
+	}
+
+	// New%(svcName) returns a new %(svcName)
+	func New%(svcName)(client *%(entClient)) *%(svcName) {
+		return &%(svcName){
+			client: client,
+		}
+	}`, tmplValues{
+		"svcName":   g.service.GoName,
+		"entClient": g.entPackage.Ident("Client"),
+	})
+}
+
+func (g *serviceGenerator) generateToProtoMapper() error {
```

Comment:
I'd use a different name for this function, like `generateConvertFunc` or `generateToProtoFunc`, since we generate here a function, and I find it clearer than the word mapper which is a bit abstract.